ewatercycle/models/pcrglobwb.py

```diff
@@ -15,94 +17,175 @@
 
-        Returns: Path to config file and work dir
-        """
-        self._setup_work_dir()
-        self._setup_config(cfg_file, input_dir, **kwargs)
-        self._start_container(input_dir, additional_input_dirs)
+    def _set_docker_image(self):
+        images = {
+            "setters": "ewatercycle/pcrg-grpc4bmi:setters",
+        }
+        self.docker_image = images[self.version]
 
-        return self.cfg_file, self.work_dir,
+    def _set_singularity_image(self):
+        # TODO auto detect sif file based on docker image and singularity dir.
+        images = {
+            "setters": "ewatercycle/pcrg-grpc4bmi:setters",
+        }
+        self.singularity_image = CFG['singularity_dir'] / images[self.version]
 
     def _setup_work_dir(self):
+        # Must exist before setting up default config
         timestamp = time.strftime("%Y%m%d_%H%M%S")
         work_dir = Path(CFG["output_dir"]) / f'pcrglobwb_{timestamp}'
         work_dir.mkdir()
-        self.work_dir = work_dir.resolve()
-        print(f"Created working directory: {work_dir}")
+        self.work_dir = work_dir.expanduser().resolve()
 
-    def _setup_config(self, cfg_file: PathLike, input_dir: PathLike, **kwargs):
-        cfg = CaseConfigParser()
-        cfg.read(cfg_file)
-        self.cfg = cfg
+    def _setup_default_config(self):
+        config_file = self.parameter_set.default_config
+        input_dir = self.parameter_set.input_dir
 
-        full_input_path = Path(input_dir).resolve()
-        cfg.set('globalOptions', 'inputDir', str(full_input_path))
+        cfg = CaseConfigParser()
+        cfg.read(config_file)
+        cfg.set('globalOptions', 'inputDir', str(input_dir))
         cfg.set('globalOptions', 'outputDir', str(self.work_dir))
 
+        self.config = cfg
+
+    def setup(  # type: ignore
+            self, **kwargs) -> Tuple[PathLike, PathLike]:
+        """Start model inside container and return config file and work dir.
+
+        Args:
+
+            - **kwargs (optional, dict): Should be passed as a dict,
+            e.g. `meteoOptions = {"temperatureNC": "era5_tas_1990_2000.nc"}`
+            where meteoOptions is the section in which the temperatureNC option
+            may be found. See :py:attr:`~parameters` for all available settings.
+
+        Returns: Path to config file and work dir
+        """
+        self._update_config(**kwargs)
+        cfg_file = self._export_config()
+        work_dir = self.work_dir
+
+        self._start_container()
+
+        return cfg_file, work_dir
+
+    def _update_config(self, **kwargs):
+        cfg = self.config
+
+        default_input_dir = self.parameter_set.input_dir
+
         for section, options in kwargs.items():
+            if not cfg.has_section(section):
+                cfg.add_section(section)
+
             for option, value in options.items():
-                cfg.set(section, option, value)
 
+                if Path(value).exists():
```

Comment:
On line 50 a value can be set to an empty string. Doing

```
In [1]: from pathlib import Path

In [4]: p = Path('')

In [5]: p.exists()
Out[5]: True

```

So when I run the notebook I get a pcrglobwb_ewatercycle.ini with
```
refETPotFileNC = /home/verhoes/git/eWaterCycle/ewatercycle/examples
```

I expected the refETPotFileNC to have kept the same value as the default config 
```
refETPotFileNC = forcing/CRU-TS3.21/merged_1958_to_2010/CRU-TS3.21_ERA-40_ERA-Interim_daily_referencePotET_1958_to_2010.nc

```

Reply:
Not sure how you triggered this, but I think it is solved now.

Reply:
By running examples/pcrglobwb.ipynb notebook until the `setup()` method. I then looked at output of `pcrglob.parameters`:

```
('meteoOptions',
  {'precipitationNC': '/home/verhoes/git/eWaterCycle/ewatercycle/examples/pcrglobwb_example_case/forcing/precipitation_2001to2010.nc',
   'temperatureNC': '/home/verhoes/git/eWaterCycle/ewatercycle/examples/temperature_2001to2010.nc',
   'referenceETPotMethod': 'Hamon',
   'refETPotFileNC': 'forcing/CRU-TS3.21/merged_1958_to_2010/CRU-TS3.21_ERA-40_ERA-Interim_daily_referencePotET_1958_to_2010.nc'}),
```

If I run the notebook again the `refETPotFileNC` value now is correct as it is unchanged from the default_config.

However the temperatureNC value is weird
Because you used 
```
forcing = ewatercycle.forcing.load_foreign(
    target_model = "pcrglobwb",
    start_time = "2001-01-01T00:00:00Z",
    end_time = "2010-12-31T00:00:00Z",
    directory = "./pcrglobwb_example_case/forcing",
    shape = None,
    forcing_info = dict(
        precipitationNC = "pcrglobwb_example_case/forcing/precipitation_2001to2010.nc",
        temperatureNC = "temperature_2001to2010.nc"
))
```

I expected value of temperatureNC to be `forcing.directory / "temperature_2001to2010.nc"` so after relativizing against `inputDir` I expected `forcing/temperature_2001to2010.nc`. However now it is set to  '/home/verhoes/git/eWaterCycle/ewatercycle/examples/temperature_2001to2010.nc' which is `./temperature_2001to2010.nc` relative to notebook.

Reply:
Solved now. The current scheme:
- User supplies `directory` and `temperatureNC` (relative to `directory`; clearly documented) in constructing the forcing object
- The model class will combine them and then expand into an absolute path
- That path is mounted on the container and substituted in the config file. 

There is still a way to configure non-default paths. `setup` now supports an "additional_dirs" argument, which allows the user to add any deviating paths in the config file, and to make sure the container is still able to find them, but that's an advanced use case that we should probably not advertise.